Gaussian: implement the 'CommonRelaxWorkChain' protocols

eimrek opened the pull request:
Hi!

Here's the Gaussian implementation. I also tested it together with CP2K for the NH3 inversion energetics (energy difference between the flat and the cone configuration) with 'moderate' protocols: both gave reasonable results (same ballpark as literature values): Gaussian: 0.182 eV, CP2K: 0.226 eV.

There's a slightly annoying thing in Gaussian (and same for ORCA): you should specify the amount of memory and number of processors in the input script. And this memory should be lower than the real reserved amount of memory (for my environment, i need to ask 1.5 GB more memory than I specify in the input script.) I parsed the `calc_engines` argument for this.

bosonie commented:
Hi Kristjan, this is great! Thanks!!
Only few questions:
1) Regarding the memory: perfect you parse it in `calc_engines` and also that you have it optional, not mandatory. If it is not specified, Gaussian uses the default 800 MB, right? However I'm not sure I understood the part saying that the specified memory should be "lower than the real reserved amount of memory" and that this "lower" seems quite a lot lower. My only concern is: with the default of 800MB, do you think that will be possible to run fine on the quantum mobile? In case not? In which units you specify the memory in the `options` of `calc_engines`?
2) It's super nice you tested NH3 energies. Could you please share the initial structures you used? You can put them in `common/data` folder. It would be great to have them in a file format that `ase` can import.

eimrek commented:
Thanks Emanuele.

I added the NH3 geometry initial guesses.

Regarding the memory, let me elaborate:

Currently, the only option (to my knowledge) to ask for memory reservation/limits through AiiDA is the `metadata.options.max_memory_kb` attribute (that can also be specified in the `calc_engines` argument). The units are `KB` as the name of the attribute reflects and there shouldn't be any problems in this regard.

I did not implement any additional inputs to workchain, but just a simple heuristic that works for my environment and only gets used when the `metadata.options.max_memory_kb` is specified. Basically it just tells gaussian to use `metadata.options.max_memory_kb- 1.5 GB` amount of memory. However, I suspect that for different environments, this heuristic would vary, so we probably should implement a way to control this more explicitly.

eimrek commented:
I did some more detailed testing on this memory issue, and looks like `ulimit -v` might be a culprit here.

In my test, i was asking Gaussian for `%mem=548MB`

If I use the direct scheduler (so runs the job on a login node), I have to ask (through`metadata.options.max_memory_kb`) for a total of `2000 MB` for the calculation to run. If I ask for `1900 MB`, i get a memory allocation error. On the direct scheduler, this amount effectively only gets passed to the `ulimit -v` command.

In case of using the LSF scheduler, and reserving memory through with the BSUB commands (units are in MB):
```
#BSUB -R "rusage[mem=685]"
#BSUB -M 685
```
everything worked well (this is a common rule online that you should ask for +25% memory compared to the Gaussian input).

Everything worked well even if I asked for less:
```
#BSUB -R "rusage[mem=500]"
#BSUB -M 500
```
as the real maximum memory use was `265 MB` (writted in stdout).

Anyways, it looks like the `ulimit -v` command is perhaps inappropriate in the direct scheduler. For sure I shouldn't always ask for extra 1.5 GB just to work with that command. What I propose is: I implement by default the "+25% heuristic" (i.e. Gaussian get ~25% less memory than the specified amount). And an optional keyword in the `calc_engines` argument, if one wants to explicitly specify the gaussian memory. What do you think?

bosonie commented:
Thanks for the clarification. And nice you reported the problem to the aiida issues.
Regarding our case, I agree to implement by default the "-25% memory" in your inputs generator and I do not see the need for a keyword to specify directly the gaussian memory. In fact, an expert user can always modify directly the `parameters` port of the `builder` before submission. Since the beginning we wanted to leave full flexibility to the users, the builder allows to do that because is not stored.
On the other side of the spectrum there is the situation when also the `calc_engines` is automatically set up (see for instance CLI of this project: https://github.com/aiidateam/aiida-common-workflows/blob/master/aiida_common_workflows/cli/launch.py) and this concerns me a bit more. Based on your experience, how likely is to encounter problems when no `max_memory_kb` is specified in calc engines (meaning, according to your implementation, gaussian uses the default 800 MB)?

bosonie commented:
One last thing. You did not implement any logic in the generator for `threshold_forces`, `threshold_stress` and `previous_workchain`. While the last two might not be as relevant for gaussian, I guess  `threshold_forces` should be a quantity the users can select in the input of gaussian, no?

eimrek commented:
Thanks for the comments Emanuele.

Regarding the memory specification: I set a default of 2 GB to be specified to Gaussian if memory is not set through `max_memory_kb` (actually the Gaussian default for gaussian09 is 256MB instead of 800MB that I mentioned before). This should be enough for most DFT-based calculations and also it's not too large that might cause any problems in the submitted system. If `max_memory_kb` is specified, then it will apply the -20% rule. I think these settings are quite robust. I can also run the calculations now with the direct scheduler by not specifying any memory (so the `ulimit -v` line is not set) and it should also work well with the automatic (CLI) set-up you were mentioning. Regarding the comment that "expert user can always modify directly the `parameters` port of the `builder`" I think this doesn't apply when you want to specify the memory to Gaussian input script (only the currently implemented heuristic will modify that) but i think this is not really a problem.

Regarding `threshold_forces`, unfortunately Gaussian doesn't really allow to set this value explicitly (at least to my knowledge). The way to specify the convergence threshold is through the directives `opt=loose`, `opt`, `opt=tight`, `opt=verytight`, . Currently I implemented this selection in the protocols: `fast` uses the default `opt`, while `moderate` and `precise` use the `opt=tight` directive. Let me know if it should be done differently.

The `previous_workchain` argument might be applicable... I was not fully sure what functionality we want there and I didn't really see the other code implementations support it either (maybe i just missed the details though).

bosonie reviewed (dismissed):
Quoting from the wiki page:
> **previous_workchain**. A `<Code>RelaxWorkChain` node. Optional. It is expected that each developer implements the logic in `get_builder` to return inputs that ensure to run a `<Code>RelaxWorkChain` that will results in energies comparable to “previous_workchain”. This is necessary to create, for instance, meaningful equations of state. The presence of this optional input, alone, trigger the process.

In case it applies to gaussian, you can added it in a separate PR, also the other codes are a bit slow in implementing it.
All the rest is great, thanks a million! Ready to be squashed and merged.

eimrek commented:
Thanks for the approve. I was wrong about the `threshold_forces`, there is a way to specify it explicitly to through the "IOp" options (this special interface to the internal parameters). I will try to implement those, I also have to check if the plugin support setting these options at the moment. But then I have to decide, what will take precedence, either e.g. `opt=tight` or specifying this threshold...

I think what makes sense is to use `threshold_forces=None` as default and if it's set, then set the internal parameter.

bosonie commented:
Yes. In any code, if no `threshold_forces` is specified, the protocol will select automatically what to do. Gaussian can implement the same, little matter that in the protocols you use `opt` and for the custom `threshold_forces` you set something else.

eimrek commented:
Ok now the `threshold_forces` is supported. Also before the `forces` output was in atomic units, fixed that as well.

There is a limitation in gaussian, that you cannot set the forces threshold to be lower than 1e-6 au = 5e-5 eV/ang. I implemented such that if the threshold is set lower, it will reset it back to 1e-6. Also I am printing a warning if this happens.

bosonie reviewed (approved):
Sorry didn't realize I needed to approve again. I will merge myself tomorrow if you will not do it. Thanks!!
